### ISSUE TITLE
Property Payload does not exist on type

### ISSUE THREAD

**Issue by bufke** [MrWolfZ/ngrx-forms#75]
See https://github.com/ngrx/platform/issues/31

There are times I'd like to monitor ngrx-form actions. For example - I handle a server side error in my own reducer instead of ngrx-forms. I'd like to hide the error when SetValueAction is run. So I can do something like this

```
  switch (action.type) {
    case SetValueAction<any>:
      return {
        ...state,
        isEmailTaken: false
      };
```

However this causes the Property Payload does not exist on type type error from the linked issue to occur. Honestly I'm losing my mind a little on what makes it work vs not work:

- If I import SetValueAction as seen in that snippet I get the property payload error on any other actions with a payload.
- If I copy and paste the source of SetValueAction into my project - it works just fine.

This sounds silly, so I must be wrong about some assumption somewhere. It seems easy enough to reproduce. It seems like the "readonly" gets lost when importing SetValueAction.

I'm using Typescript 2.7.2 with Angular 6.

**Comment by bufke** [comment]
Sorry yes that was a typo. A better example

```javascript
export class MyAction implements Action {
  type: 'myAction';
  a: string;
}

export class MySecondAction implements Action {
  type: 'mySecondAction';
  b: number;
}

export type MyActions = MyAction | MySecondAction;

export function badReducer(
  state = {},
  action: MyActions | SetValueAction<any>,
) {
  switch (action.type) {
    case: 'myAction':
      // TypeScript is expected to say action is MyAction
      // TypeScript instead says action is MyAction | MySecondAction | SetValueAction
      return state;
    case: SetValueAction.TYPE:
      // TypeScript says action is MyAction | MySecondAction | SetValueAction
      return state;
  }
}

export function goodReducer(
  state = {},
  action: MyActions,
) {
  switch (action.type) {
    case: 'myAction':
      // TypeScript says action is MyAction
      return state;
}
```

**Comment by MrWolfZ** [comment]
Hmm, indeed, the type definitions are wrong. They do not contain the proper literal values for the `type` property. At first I though this must be a regression I introduced with a recent change but this behaviour exists at least since 2.0.0 (didn't check further in the past). Curious that I never encountered this issue myself before.

For now I can only offer a very ugly workaround. Simply declare the proper values yourself. I will look into a fix for this.

```typescript
// reducer.ts
import { SetValueAction } from 'ngrx-forms';
import { MyActionsUnion, MyActionTypes } from './actions';

interface State {
  foo: string;
}

const initialState: State = {
  foo: '',
};

const SET_VALUE_TYPE = 'ngrx/forms/SET_VALUE';
interface FixedSetValueAction<TValue> extends SetValueAction<TValue> {
  readonly type: typeof SET_VALUE_TYPE;
}

export function reducer(state = initialState, action: MyActionsUnion | FixedSetValueAction<any>) {
  switch (action.type) {
    case MyActionTypes.MyAction:
      console.log(action.payload); // <<< no TS ERROR!
      return state;
    case MyActionTypes.MySecondAction:
      console.log('no payload', action);
      return state;
    case SET_VALUE_TYPE:
      console.log(action.payload);
      return state;
    default:
      return state;
  }
}
```

**Comment by MrWolfZ** [comment]
Alright, the fix was easy enough. I have just published version 2.3.5 which contains a fix for this issue.
